base/show.jl

```diff
@@ -487,37 +484,256 @@ function io_has_tvar_name(io::IOContext, name::Symbol, @nospecialize(x))
             end
         end
-        print(io, "Union")
-        show_delim_array(io, uniontypes(x), '{', ',', '}', false)
-        return
     end
-    x::UnionAll
+    if x isa Union
+        y = []
+        normal = true
+        for typ in uniontypes(x)
+            if isa(typ, TypeVar)
+                normal = false
+            else
+                push!(y, typ)
+            end
+        end
+        normal || (x = Union{y...})
+        properx = rewrap_unionall(x, properx)
+    end
+    has_free_typevars(properx) && return Any
+    return properx
+end
+
+function make_typealias(x::Type)
+    Any <: x && return
+    x <: Tuple && return
+    mods = modulesof!(Set(), x)
+    Core in mods && push!(mods, Base)
+    aliases = Tuple{GlobalRef,SimpleVector}[]
+    xenv = UnionAll[]
+    for p in uniontypes(unwrap_unionall(x))
+        p isa UnionAll && push!(xenv, p)
+    end
+    x isa UnionAll && push!(xenv, x)
+    for mod in mods
+        for name in names(mod)
+            if isdefined(mod, name) && !isdeprecated(mod, name) && isconst(mod, name)
+                alias = getfield(mod, name)
+                if alias isa Type && !has_free_typevars(alias) && !isvarargtype(alias) && !print_without_params(alias) && x <: alias
+                    if alias isa UnionAll
+                        (ti, env) = ccall(:jl_type_intersection_with_env, Any, (Any, Any), x, alias)::SimpleVector
+                        # ti === Union{} && continue # impossible, since we already checked that x <: alias
+                        env = env::SimpleVector
+                        # TODO: In some cases (such as the following), the `env` is over-approximated.
+                        #       We'd like to disable `fix_inferred_var_bound` since we'll already do that fix-up here.
+                        #       (or detect and reverse the compution of it here).
+                        #   T = Array{Array{T,1}, 1} where T
+                        #   (ti, env) = ccall(:jl_type_intersection_with_env, Any, (Any, Any), T, Vector)
+                        #   env[1].ub.var == T.var
+                        applied = alias{env...}
+                        for p in xenv
+                            applied = rewrap_unionall(applied, p)
+                        end
+                        has_free_typevars(applied) && continue
+                        applied == x || continue # it couldn't figure out the parameter matching
+                    elseif alias <: x
+                        env = Core.svec()
+                    else
+                        continue # not a complete match
+                    end
+                    push!(aliases, (GlobalRef(mod, name), env))
+                end
+            end
+        end
+    end
+    if length(aliases) == 1 # TODO: select the type with the "best" (shortest?) environment
+        return aliases[1]
+    end
+end
+
+function show_typealias(io::IO, name::GlobalRef, x::Type, env::SimpleVector)
+    # print(x.mod)
```

Comment:
I suppose this should use the standard logic on when to print qualified names?

Reply:
Yes, that's a good TODO item I forgot about. Curiously enough though, if I wasn't specifically filtering them out, this here would nearly subsume that logic. (since it's computing the dual of the "is there an existing exported alias of the appropriate name" logic that we're doing here)